core/charm/baseselector.go

```diff
@@ -0,0 +1,194 @@
+// be deployed to.
+type BaseSelector struct {
+	requestedBase       series.Base
+	defaultBase         series.Base
+	explicitDefaultBase bool
+	force               bool
+	logger              SelectorLogger
+	supportedBases      []series.Base
+	usingImageID        bool
+}
+
+type SelectorConfig struct {
+	Config              SelectorModelConfig
+	Force               bool
+	Logger              SelectorLogger
+	RequestedBase       series.Base
+	SupportedCharmBases []series.Base
+	// usingImageID is true when the user is using the image-id constraint
+	// when deploying the charm. This is needed to validate that in that
+	// case the user is also explicitly providing a base.
+	UsingImageID bool
+}
+
+type SelectorModelConfig interface {
+	// DefaultBase returns the configured default base
+	// for the environment, and whether the default base was
+	// explicitly configured on the environment.
+	DefaultBase() (string, bool)
+
+	// ImageStream returns the simplestreams stream
+	// used to identify which image ids to search
+	// when starting an instance.
+	ImageStream() string
+}
+
+// ConfigureBaseSelector returns a configured and validated BaseSelector
+func ConfigureBaseSelector(cfg SelectorConfig) (BaseSelector, error) {
+	// TODO (hml) 2023-05-16
+	// Is there more we can do here and reduce the prep work
+	// necessary for the callers?
+	imageStream := cfg.Config.ImageStream()
+	workloadBases, err := series.WorkloadBases(jujuclock.WallClock.Now(), cfg.RequestedBase, imageStream)
+	if err != nil {
+		return BaseSelector{}, errors.Trace(err)
+	}
+	defaultBase, explicit := cfg.Config.DefaultBase()
+	var parsedDefaultBase series.Base
+	if explicit {
+		parsedDefaultBase, err = series.ParseBaseFromString(defaultBase)
+		if err != nil {
+			return BaseSelector{}, errors.Trace(err)
+		}
+	}
+	bs := BaseSelector{
+		requestedBase:       cfg.RequestedBase,
+		defaultBase:         parsedDefaultBase,
+		explicitDefaultBase: explicit,
+		force:               cfg.Force,
+		logger:              cfg.Logger,
+		usingImageID:        cfg.UsingImageID,
+	}
+	bs.supportedBases, err = bs.validate(cfg.SupportedCharmBases, workloadBases)
+	if err != nil {
+		return BaseSelector{}, errors.Trace(err)
+	}
+	return bs, nil
+}
+
+// TODO(nvinuesa): The force flag is only valid if the requestedBase is specified
+// or to force the deploy of a LXD profile that doesn't pass validation, this
+// should be added to these validation checks.
+func (s BaseSelector) validate(supportedCharmBases, supportedJujuBases []series.Base) ([]series.Base, error) {
+	// If the image-id constraint is provided then base must be explicitly
+	// provided either by flag either by model-config default base.
+	if s.logger == nil {
+		return nil, errors.NotValidf("empty Logger")
+	}
+	if s.usingImageID && s.requestedBase.Empty() && !s.explicitDefaultBase {
+		return nil, errors.Forbiddenf("base must be explicitly provided when image-id constraint is used")
+	}
```

Comment:
This feels like a cross-cutting concen. I wonder if this check should have been performed somewhere else, not necessarily inside the `BaseSelector`. Maybe we should have another validator. I've got a feeling we're just shoving too much in here. It's leaving me with a bad taste.

Reply:
Agreed and we are. The BaseSelector needs more changes, I'm just running out of time and scope right now. It's only being used in one place behind a feature flag. The next step is to polish it and use it everywhere. Which I am passing to @jack-w-shaw.